src/Datadog.Trace/ScopeManager.cs

```diff
@@ -0,0 +1,143 @@
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            Log.ErrorException($"Active scope access failed: {_prioritizedScopeAccess[i]?.GetType().FullName}.", ex);
+                            continue;
+                        }
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        public Scope Activate(Span span, bool finishOnClose)
+        {
+            var newParent = Active;
+            var scope = new Scope(newParent, span, this, finishOnClose);
+            var scopeOpenedArgs = new SpanEventArgs(span);
+
+            SpanOpened?.Invoke(this, scopeOpenedArgs);
+
+            SetActiveScope(scope);
+
+            if (newParent != null)
+            {
+                SpanDeactivated?.Invoke(this, new SpanEventArgs(newParent.Span));
+            }
+
+            SpanActivated?.Invoke(this, scopeOpenedArgs);
+
+            return scope;
+        }
+
+        public void Close(Scope scope)
+        {
+            var current = Active;
+            var isRootSpan = scope.Parent == null;
+
+            if (current == null || current != scope)
+            {
+                // This is not the current scope for this context, bail out
+                return;
+            }
+
+            // if the scope that was just closed was the active scope,
+            // set its parent as the new active scope
+            SetActiveScope(current.Parent);
+            SpanDeactivated?.Invoke(this, new SpanEventArgs(current.Span));
+
+            if (!isRootSpan)
+            {
+                SpanActivated?.Invoke(this, new SpanEventArgs(current.Parent.Span));
+            }
+
+            SpanClosed?.Invoke(this, new SpanEventArgs(scope.Span));
+
+            if (isRootSpan)
+            {
+                TraceEnded?.Invoke(this, new SpanEventArgs(scope.Span));
+            }
+        }
+
+        public void RegisterScopeAccess(IActiveScopeAccess scopeAccess)
+        {
+            lock (_scopeAccessLock)
+            {
+                if (_prioritizedScopeAccess.Any(i => i.GetType() == scopeAccess.GetType()))
+                {
+                    // We don't want multiple instances as each type manages scope in an ambient fashion.
+                    return;
+                }
+
+                _prioritizedScopeAccess.Add(scopeAccess);
+
+                // Slower add for faster read
+                _prioritizedScopeAccess = _prioritizedScopeAccess.OrderByDescending(i => i.Priority).ToList();
+            }
+        }
+
+        private void SetActiveScope(Scope scope)
+        {
+            lock (_scopeAccessLock)
+            {
+                for (int i = 0; i < _prioritizedScopeAccess.Count; i++)
+                {
+                    try
+                    {
+                        if (_prioritizedScopeAccess[i].TrySetActiveScope(scope))
```

Comment:
Since the list of `IActiveScopeAccess` is ordered by descending priority, I think that ASP.NET apps will use `AsyncLocalCompatScopeAccess)` first, return `true`, and never call `AspNetActiveScopeAccess.TrySetActiveScope()`. 

Maybe `AspNetActiveScopeAccess` should have higher priority than `AsyncLocalCompatScopeAccess` (it's only added to the list for ASP.NET apps). Or maybe we don't need priority and we call `TrySetActiveScope()` on all items on the list.

Reply:
I'm not sure what I was thinking in the moment, it should just be OrderBy, higher number being higher priority.

I was thinking we'd only use highest priority in pretty much all cases, and the others would be used as fallbacks if we were in some weird context, like maybe a background thread.

I didn't know if considering "promotion" logic was a reasonable thing to do yet.
i.e., if a lower priority instance has a non-closed scope, promote it to the higher priority ones if possible. Maybe that's a bit too much to do right now.

Reply:
I like Lucas's idea where `TrySetActiveScope()` sets the scope on all of the `IActiveScopeAccess` objects. It would just be the getter that would get the highest priority one. Would there be downsides to that approach?

Reply:
Possible performance and memory implications.
Some tracers report many tens of thousands of spans per trace.
We'd be storing that several times over potentially.

Reply:
If we move toward flushing spans as they complete rather than when the trace completes, I think I am all for this.